tests/base/src/test/java/org/wildfly/security/ssl/SSLv2HelloAuthenticationTest.java

```diff
@@ -0,0 +1,444 @@
+                "as IBM JDK does not support SSLv2Hello on the " +
+                "client side", IS_IBM);
+
+        List<Protocol> list = new ArrayList<>();
+        list.add(Protocol.forName("SSLv2Hello"));
+        list.add(Protocol.forName("TLSv1"));
+
+        SSLContext serverContext = new SSLContextBuilder()
+                .setSecurityDomain(securityDomain)
+                .setKeyManager(getKeyManager("/ca/jks/scarab.keystore"))
+                .setTrustManager(getCATrustManager())
+                .setNeedClientAuth(true)
+                .setProtocolSelector(ProtocolSelector.empty().add(EnumSet.copyOf(list)))
+                .build().create();
+
+        String[] enabledProtocols = new String[]{"SSLv2Hello", "TLSv1"};
+
+        SecurityIdentity identity = performConnectionTest(serverContext,
+                "protocol://test-two-way-sslv2hello.org",
+                "wildfly-ssl-test-config-v1_6.xml",
+                enabledProtocols, // We expect client and server socket to only have SSLv2Hello and TLSv1 enabled
+                "TLSv1"); // We expect the negotiated protocol to be TLSv1, as SSLv2Hello is a pseudo-protocol
+
+
+        assertNotNull(identity);
+        assertEquals("Principal Name", "ladybird", identity.getPrincipal().getName());
+    }
+
+    /**
+     * Verify SSLv2Hello is not enabled in client or server if not explicitly configured
+     */
+    @Test
+    public void testTwoWaySSLv2HelloNotEnabled() throws Exception {
+        List<Protocol> list = new ArrayList<>();
+        list.add(Protocol.forName("TLSv1"));
+        list.add(Protocol.forName("TLSv1.1"));
+
+        SSLContext serverContext = new SSLContextBuilder()
+                .setSecurityDomain(securityDomain)
+                .setKeyManager(getKeyManager("/ca/jks/scarab.keystore"))
+                .setTrustManager(getCATrustManager())
+                .setNeedClientAuth(true)
+                .setProtocolSelector(ProtocolSelector.empty().add(EnumSet.copyOf(list)))
+                .build().create();
+
+        String[] enabledProtocols = new String[]{"TLSv1", "TLSv1.1"};
+
+        SecurityIdentity identity = performConnectionTest(serverContext,
+                "protocol://two-way-no-sslv2hello.org",
+                "wildfly-ssl-test-config-v1_6.xml",
+                enabledProtocols, // We expect only TLSv1 and TLSv1.1 to be enabled i.e. SSLv2Hello should only be enabled if explicitly configured
+                "TLSv1.1"); // We expect the negotiated protocol to be the highest version protocol in common
+
+        assertNotNull(identity);
+        assertEquals("Principal Name", "ladybird", identity.getPrincipal().getName());
+    }
+
+    /**
+     * Test two way authentication when the client does not support SSLv2Hello,
+     * but the server has SSLv2Hello support enabled.
+     */
+    @Test
+    public void testTwoWaySSLv2HelloNoClientSupport() throws Exception {
```

Comment:
Would also be good to test the case where SSLv2Hello is not enabled on the server side but it is enabled on the client side.

Reply:
added